lib/temporal/activity/context.rb

```diff
@@ -39,8 +41,29 @@ def async_token
         )
       end
 
+      # Returns true if the activity has been canceled directly by its workflow or indirectly by
+      # its workflow being terminated. This will only be set following a call to heartbeat that is
+      # not throttled.
+      def cancel_requested
+        @cancel_requested
+      end
+
+      # Returns true if the activity has breached its start-to-close or schedule-to-close timeout
+      def timed_out?
+        deadline_exceeded?(start_to_close_deadline) || deadline_exceeded?(schedule_to_close_deadline)
+      end
+
+      # This returns true if the worker has started shutting down upon receiving a
+      # TERM or INT signal. Once this happens, your activity should finishing processing
+      # quickly or raise an error to fail the activity attempt.
+      def shutting_down?
```

Comment:
In the new SDK we've taken a slightly different approach and I wonder if it might be a good idea to steer this into the same direction. Basically instead of asking whether the worker is shutting down, the worker would attempt to cancel the activity (similar process to activity cancellation, except using a different exception).

The main reason here is to make sure that cancellations are explicitly handled (including explicitly ignored). By exposing the `shutting_down?` method we're still defaulting to unhandled behaviour and I assume most activities won't use it at all.

Reply:
I'd be curious to learn more about the API and mechanics of this. Is the idea that the worker propagates the cancellation request to the activity code and the activity then acks the cancellation once it's ready to shutdown? How is it different than simply raising an error that can be rescued and re-raised later?

Reply:
I spent some time looking through the sdk-ruby code. This is much clearer to me now. I agree this pattern should work better.

However, I do have concerns about changing the behavior of the existing API where there are currently no errors raised either directly by the heartbeat method or through thread interruption. For now, here's how I'm thinking to deal with this:
- Add these methods onto the activity context:
  - `cancel` for communicating cancellation reasons via an error argument
  - `cancelled?` an alias to `cancel_requested` that returns a Boolean. This can be checked by activities that don't want to deal with errors being raised by `heartbeat!` or who don't want to heartbeat at all, but rather check for cancellation at certain critical points of the activity.
- Use the `cancel` method to communicate cancellation to the context for the three cases in this PR. This should help clean up some icky code like `is_shutting_down_proc`
- Eliminate `ActivityInterrupted`, instead making `ActivityCanceled` the base class for naming consistency. Having subclassed errors is still useful, but I'm open to collapsing these into a single error like in sdk-ruby too.
- For now, instead of raising an error onto the thread, this errors will only be raised out of the `heartbeat_interrupted`/`heartbeat!` method added in this PR already
- Eventually, an activity or worker option can be added for thread interruption behavior. A `shield` method can also be written to protect certain sections of activity code from being interrupted.